### PR TITLE
Fix Issue 20868 - DIP1000: scope delegate triggers error in unsafe co…

### PR DESCRIPTION
**Pull request by RazvanN7**
…de and it shouldn't

cc @WalterBright @AndrejMitrovic

### CODE REVIEW

File: test/compilable/test20868.d [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// https://issues.dlang.org/show_bug.cgi?id=20868
+// REQUIRED_ARGS: -dip1000
+
+void scoped (scope void delegate() dg)
+{
+    static void delegate()[] dgs;
+    dgs ~= dg;  // error
```
Comment (MoonlightSentinel) [inline test/compilable/test20868.d:7]: Shouldn't `scope` apply to the context pointer?

`scope` checks are also not restricted to `@safe`, see e.g. 

```d
int* scoped(scope int* p) {
    return p;
}

void main() {
    int i;
    int* p = &i;
    scoped(p);
}
```
Compiling with `-dip1000`
```
onlineapp.d(2): Error: scope variable `p` may not be returned
```

Comment (RazvanN7) [reply test/compilable/test20868.d:7]: I'm not sure. Spec [1] says: "scope escape analysis is only done for @safe functions. For other functions scope semantics must be manually enforced.". Dip1000 [2] says: "Errors for scope violations are only reported in @safe code.".

However, there are cases like the one that you have highlighted where scope is enforced even if not in a @safe function.

[1] https://dlang.org/spec/function.html#scope-parameters
[2] https://github.com/dlang/DIPs/blob/master/DIPs/other/DIP1000.md

**Comment by thewilsonator** [comment]
The error in the issue is passing a `scope delegate` to a function that accepts a `delegate`. The test case here is wrong, the test should error.

File: test/compilable/test20868.d (hunk 1 above)
Comment (MoonlightSentinel) [reply test/compilable/test20868.d:7]: Guess i dind't remember [1] (and [2] is outdated anyways).
But storage classes depending on function attributes doesn't seem right.

**Review by thewilsonator — DISMISSED** [review]
See comment

File: test/compilable/test20868.d (hunk 1 above)
Comment (RazvanN7) [reply test/compilable/test20868.d:7]: @thewilsonator I am not sure if your review is right. Check the above comments. That is why we need some clarifications from @WalterBright .

Comment (thewilsonator) [reply test/compilable/test20868.d:7]: This test case being incorrect (i.e. compilable) has nothing to do with dip1000 and has everything to do with the original purpose of `scope delegate`s:

`scope` on a delegate means "this argument's context pointer does not outlive a call to this function, do not GC allocate a closure for it"

Comment (RazvanN7) [reply test/compilable/test20868.d:7]: Just to clarify, here is the original bug report:

```d
void scoped (scope void delegate() dg)
{
    static void delegate()[] dgs;
    dgs ~= dg;  // error

    nonScoped(dg);  // not an error???
}

void nonScoped (void delegate() dg)
{
    static void delegate()[] dgs;
    dgs ~= dg;
}

void main ()
{
    int x;
    scoped({x = 100;});
}
```

Without this patch:

1. If you compile without @safe or dip1000, you get no error.
2. If you compile with dip1000, you get an error at the line that has the error comment
3. If you compile with dip1000 and annotate the functions with @safe you get errors on both the line that has "error" comment and the line that has "non an error???" comment.

So, what is the correct behavior here? As far as the spec goes, scope should do nothing without -dip1000 and @safe, however, the language implementation sometimes does and sometimes doesn't respect that. So I am confused on what the fix should be or if there should be a fix at all.

If we consider the spec as the universal truth, then we should patch the compiler to ignore scope whenever we are not in a @safe function (non-dip1000 context).

@thewilsonator What is your source of truth?

**Comment by RazvanN7** [comment]
@thewilsonator See the comments on the bug report.

**Review by Geod24 — APPROVED** [review]
I really think it's a mistake that `-preview=dip1000` does not apply to non-`@safe` code.
But in the meantime, this fix only affects a `-preview` switch, and make the compiler behaves according to the spec, so LGTM.

**Comment by thewilsonator** [comment]
As I said this has nothing to do with `@safe` and everything to do with the prior behaviour of scope delegates. Please don't mark reviews as stale when they haven't been addressed.

**Pull request merged**